File: express-backend/src/models/Blob.ts
import mongoose, { Schema, type Document, type Model } from 'mongoose';

export interface BlobDocument extends Document {
  storage_backend: string;
  storage_key: string;
  filename: string;
  content_type: string;
  size_bytes: number;
  uploaded_by?: mongoose.Types.ObjectId;
  created_at: Date;
}

const BlobSchema = new Schema<BlobDocument>(
  {
    storage_backend: { type: String, required: true },
    storage_key: { type: String, required: true },
    filename: { type: String, required: true },
    content_type: { type: String, required: true },
    size_bytes: { type: Number, required: true },
    uploaded_by: { type: Schema.Types.ObjectId, ref: 'User' },
    created_at: { type: Date, default: () => new Date() },
  },
  {
    collection: 'blobs',
  },
);

export const BlobModel: Model<BlobDocument> =
  mongoose.models.Blob || mongoose.model<BlobDocument>('Blob', BlobSchema);

File: backend/app/schemas/blob.py
"""Blob (file) metadata document schema. Actual bytes in GridFS/S3/local."""
from datetime import datetime, timezone
from typing import Optional

from pydantic import BaseModel, Field

from .enums import StorageBackend
from .objectid import PyObjectId


def _utc_now() -> datetime:
    return datetime.now(timezone.utc)


class BlobDoc(BaseModel):
    """Mongo document shape for blobs collection. References stored in messages.attachment_ids."""

    id: Optional[PyObjectId] = Field(None, alias="_id")
    storage_backend: StorageBackend
    storage_key: str
    filename: str
    content_type: str
    size_bytes: int
    uploaded_by: Optional[PyObjectId] = None
    created_at: datetime = Field(default_factory=_utc_now)

    model_config = {"populate_by_name": True, "arbitrary_types_allowed": True}
